framework/source/class/qx/core/MObjectId.js

```diff
@@ -157,59 +166,85 @@ qx.Mixin.define("qx.core.MObjectId", {
         if (found) {
           result = target;
         }
-        
+
       } else {
         // No object, creating the object
         result = this._createQxObject(id);
       }
-      
+
       if (result && controlId) {
         var childControl = result.getChildControl(controlId);
         return childControl;
       }
-      
+
       return result;
     },
-    
+
     /**
      * Creates the object and adds it to a list; most classes are expected to
-     * override `_createQxObjectImpl` NOT this method.
-     * 
+     * override `_createQxObjectImpl` NOT this method. If `_createQxObjectImpl`
+     * returns `undefined`, this is almost always the result of a misconfiguration
+     * (such as a typo in the id), since `null` is used if you explicitly do
+     * not want to return a value from `_createQxObjectImpl`. To address
+     * this, a warning is emitted if the environment variable `"qx.debug"` is
+     * `true`. In order to make the source of the problem easier to locate,
+     * set the environment variable `"module.objectId.debugger”`
+     * to `true` to trigger the browser debugger.
+     *
      * @param id {String} ID of the object
+     * @param owner {qx.core.Object?} If given, the owner object to which the
+     * created object is attached. Otherwise, the current instance (`this`) is used.
      * @return {qx.core.Object?} the created object
      */
-    _createQxObject: function(id) {
-      var result = this._createQxObjectImpl(id);
+    _createQxObject: function(id, owner) {
+      if (owner === undefined) {
+        owner = this;
+      } else if (!(owner instanceof qx.core.Object)) {
+        throw new Error("Owner must be instance of qx.core.Object or undefined.");
+      }
+      var result;
+      if (typeof this.createQxObject == "function") {
+        // more user-friendly method name
+        result = this.createQxObject(id);
+      } else {
+        // method name of the original implementation
+        result = this._createQxObjectImpl(id);
+      }
       if (result !== undefined) {
-        this.addOwnedQxObject(result, id);
+        owner.addOwnedQxObject(result, id);
```

Comment:
Sorry but as it stands I'm not so keen on this either - although we have built it so that it is possible to add objects by ID, that's really intended for situations where the object does not implement `_createQxObjectImpl`, ie it's a work around for existing code.  The workaround means that objects can be located via a global ID string (eg "dlgLoginForm/edtUsername"), the initial use case was for automated testing.

The coding pattern which is so useful is centered around an object creating it's own objects, and they are exposed kind of like an API.  For example, a login form dialog box might expose objects with the IDs "edtUsername", "edtPassword", and "btnOk" and document them as two text fields and a button; there will be a lot more widgets used to make up that form but these are not necessarily exposed (or possibly the IDs could be prefixed with "_" or "__").  

Preserving the Owner/Ownee relationship is critical because it allows name conflicts to be avoided - but when we promote adding objects to another objects, we're breaking that protection because we are adding the new object directly into the namespace of the `owner`.  

This could have some use, in that you would be able to override the creation of a widget (eg provide the `edtUsername` textfield of a login dialog) **but** this will only work if the object has not yet been created by the dialog.  If the dialog is later modified in such a way that the constructor calls `getQxObject("edtUsername")`, then the calling code will immediately break.  The correct, and the only reliable, way to change the implementation of a field would be to derive a class from the login dialog and provide your own implementation of `_createQxObjectImpl`.

Also, if we consider the IDs an API, then the object has the right to define new objects and change or remove protected/private objects as needed; but this solution will add additional, owned objects into the namespace of an object on the fly.  Even putting name conflicts to one side for a moment, I'm not convinced that this is the right way either - it feels quite unstructured, and the equivalent of adding methods to an object on the fly.  I feel that the expected way to add functionality would be by deriving a class, and that would mean that the additions form part of the code structure (as opposed to runtime) and are much better understood and documented - you could for example easily detect all classes which derive and then check their `_createQxObjectImpl` and then know *for a fact* what API object IDs are valid in that object.

AIUI the intention of this change is to support compositing of objects, ie assist building the object graph - but I think that the correct way to do that is to either use the owner/ownee ID mechanism, where you can have "grand-owner/owner/ownee/etc" or to have a class which is specifically intended for compositing in that way.  That special "compositing" class would be the place to support a lot of other features which are very relevant to compositing (such as recording modified state, persistence, etc).

Reply:
Thanks for the comments and I see your points. I'll change the PR (or what's left of it) later. Good to get the discussion going!